### ISSUE TITLE
Unable to connect to database started with swarm (previously working in v3.3.0)

### ISSUE THREAD

**Issue by chingc** [docker/for-mac#5655]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 67DACEEA-B4D0-45A8-8F43-DB05885AA531/20210505160255

### Expected behavior

User can connect to a postgres database that was started by a docker-compose.yml file using docker swarm.

### Actual behavior

Unable to do so.

### Information

- Is it reproducible? Yes
- Is the problem new? Yes
- Did the problem appear with an update? Yes, it started with 3.3.1 and persists in 3.3.2. The last version that worked was 3.3.0.
- macOS Version: Big Sur 11.3.1
- Intel chip or Apple chip: Intel
- Docker Desktop Version: 3.3.2

### Steps to reproduce the behavior

This is the smallest docker-compose.yml file I can make to reproduce the problem:

```
version: '3.8'
services:
  postgres:
    image: postgres:latest
    environment:
      - POSTGRES_PASSWORD=12345
    ports:
      - "5432:5432"
```

Here is some of the output:

```
% docker stack deploy --compose-file docker/database/test-compose.yml pg_test
Creating network pg_test_default
Creating service pg_test_postgres

% docker stack services pg_test
ID             NAME               MODE         REPLICAS   IMAGE             PORTS
ujlodcm09jfy   pg_test_postgres   replicated   1/1        postgres:latest   *:5432->5432/tcp

% docker ps
CONTAINER ID   IMAGE             COMMAND                  CREATED         STATUS         PORTS      NAMES
12af5b98031f   postgres:latest   "docker-entrypoint.s…"   4 minutes ago   Up 4 minutes   5432/tcp   pg_test_postgres.1.b5vq3vfgbgewbt784l4ibarvz

% psql -h localhost -d postgres -U postgres -l
psql: error: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

As you can see, attempting to connect to postgres fails.  However, if I downgrade to v3.3.0, I can connect successfully and it gives the following output:

```
% psql -h localhost -d postgres -U postgres -l
Password for user postgres:
                                 List of databases
   Name    |  Owner   | Encoding |  Collate   |   Ctype    |   Access privileges
-----------+----------+----------+------------+------------+-----------------------
 postgres  | postgres | UTF8     | en_US.utf8 | en_US.utf8 |
 template0 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
(3 rows)
```

**Comment by stephen-turner** [comment]
Interesting, I can't reproduce this. I see all the thumbs up, but it's working fine for me on 3.3.1.

What happens if you use `docker-compose up` instead of `docker stack deploy`?

**Comment by chingc** [comment]
@stephen-turner Thanks for the quick reply.  I tested `docker compose up` with the newly released v3.3.3 and it worked!  However, `docker stack deploy` still fails with the same error as described in my initial post.

One thing I found interesting was when I ran `docker port` on the container started by `docker compose up`:

```
% docker port 58aa3fe48dfe
5432/tcp -> 0.0.0.0:5432
5432/tcp -> :::5432
%
```

But when I run `docker port` on the container started by `docker swarm deploy` I get nothing:

```
% docker port d205822d129d
%
```
